MMCore/MMCore.cpp

```diff
@@ -8032,6 +8042,777 @@ std::vector<std::string> CMMCore::getLoadedPeripheralDevices(const char* hubLabe
+ * After this command we can still use getImage() if we want to display the image.
+ * 
+ * \param handle - currently open dataset
+ * \param imageMeta - image metadata that we wish to add to the image
+ */
+void CMMCore::snapAndAppendToDataset(const char* handle, const std::vector<long>& coordinates, const char* imageMeta)
+{
+   snapImage();
+
+   // TODO: at this point another thread can do something with the camera
+   // and disrupt the function
+
+   std::shared_ptr<CameraInstance> camera = currentCameraDevice_.lock();
+   try {
+      // get image
+      mm::DeviceModuleLockGuard guard(camera);
+      auto pBuf = const_cast<unsigned char*> (camera->GetImageBuffer());
+      if (!pBuf)
+      {
+         logError("CMMCore::getImage()", getCoreErrorText(MMERR_CameraBufferReadFailed).c_str());
+         throw CMMError(getCoreErrorText(MMERR_CameraBufferReadFailed).c_str(), MMERR_CameraBufferReadFailed);
+      }
+
+      // process the image
+      std::shared_ptr<ImageProcessorInstance> imageProcessor = currentImageProcessor_.lock();
+      if (imageProcessor)
+      {
+         imageProcessor->Process((unsigned char*)pBuf, camera->GetImageWidth(), camera->GetImageHeight(), camera->GetImageBytesPerPixel());
+      }
+
+      // store the image
+      std::shared_ptr<StorageInstance> storage = currentStorage_.lock();
+      if (storage)
+      {
+         mm::DeviceModuleLockGuard storageGuard(storage);
+         int ret(0);
+         int imageSize = camera->GetImageWidth() * camera->GetImageHeight() * camera->GetImageBytesPerPixel();
+
+         if (coordinates.empty())
+         {
+            ret = storage->AppendImage(handle, imageSize, pBuf, imageMeta);
+         }
+         else
+         {
+            std::vector<int> coords(coordinates.begin(), coordinates.end());
+            ret = storage->AddImage(handle, imageSize, pBuf, coords, imageMeta);
+         }
+         if (ret != DEVICE_OK)
+         {
+            logError(getDeviceName(storage).c_str(), getDeviceErrorText(ret, storage).c_str());
+            throw CMMError(getDeviceErrorText(ret, storage).c_str(), MMERR_DEVICE_GENERIC);
+         }
+      }
+      else
+         throw CMMError(getCoreErrorText(MMERR_StorageNotAvailable).c_str(), MMERR_StorageNotAvailable);
+
+   }
+   catch (CMMError& e) {
+      throw e;
+   }
+   catch (...) {
+      logError("CMMCore::snapAndSave()", getCoreErrorText(MMERR_UnhandledException).c_str());
+      throw CMMError(getCoreErrorText(MMERR_UnhandledException).c_str(), MMERR_UnhandledException);
+   }
+}
+
+/**
+ * Pops the next image from the circular buffer and saves it to the dataset.
+ * This is the counterpart to popNextImage(). Instead of fetching image pixels it
+ * it sends image directly to the storage.
+ * 
+ * \param handle - currently open dataset handle
+ * \param imageMeta - image metadata
+ */
+void CMMCore::appendNextToDataset(const char* handle, const std::vector<long>& coordinates, const char* imageMeta) throw (CMMError)
```

Comment:
Possibly name `popNext[Image]AndAppendToDataset()`, to be easily compared with the plain `popNextImage()` and `appendImageToDataset()`?

---

- This won't work for `Multi Camera`. That can be addressed later (possibly with Henry's per-camera buffering).
- Applications won't be able to store metadata values generated by cameras using this mechanism. This also can be addressed later if found necessary.

(Just noting the limitations so we are aware of them.)